tools/cli/commands/connect.py

```diff
@@ -0,0 +1,263 @@
+
+    $ {0} {1} example-instance --zone us-central1-a"""
+)
+
+
+def flags(parser):
+    """Add command line flags for the `connect` subcommand.
+
+    Args:
+      parser: The argparse parser to which to add the flags.
+    """
+    parser.add_argument(
+        'instance',
+        metavar='INSTANCE',
+        help='name of the instance to which to connect')
+    connection_flags(parser)
+
+def connection_flags(parser):
+    """Add flags common to every connection-establishing subcommand.
+
+    Args:
+      parser: The argparse parser to which to add the flags.
+    """
+    parser.add_argument(
+        '--quiet',
+        dest='quiet',
+        action='store_true',
+        help='do not issue any interactive prompts')
+    parser.add_argument(
+        '--port',
+        dest='port',
+        type=int,
+        default=8081,
+        help='local port from which to forward traffic to Datalab')
+    parser.add_argument(
+        '--max-reconnects',
+        dest='max_reconnects',
+        type=int,
+        default=-1,
+        help=(
+            'maximum number of times to reconnect.'
+            '\n\n'
+            'A negative value means no limit.'))
+
+    browser_group = parser.add_mutually_exclusive_group()
+    browser_group.add_argument(
+        '--no-launch-browser',
+        dest='launch_browser',
+        action='store_false',
+        help='do not open a browser connected to Datalab')
+    browser_group.add_argument(
+        '--launch-browser',
+        dest='launch_browser',
+        action='store_true',
+        default=True,
+        help='open a browser connected to Datalab')
+
+    return
+
+
+def connect(args, gcloud_compute):
+    """Create a persistent connection to a Datalab instance.
+
+    Args:
+      args: The Namespace object constructed by argparse
+      gcloud_compute: A function that can be called to invoke `gcloud compute`
+    """
+    instance = args.instance
+    print('Connecting to {0}'.format(instance))
+
+    datalab_address = 'http://localhost:{0}/'.format(str(args.port))
+
+    def create_tunnel():
+        """Create an SSH tunnel to the Datalab instance.
+
+        This method blocks for as long as the connection is open.
+
+        Raises:
+          KeyboardInterrupt: When the end user kills the connection
+          subprocess.CalledProcessError: If the connection dies on its own
+        """
+        cmd = ['ssh']
+        if args.quiet:
+            cmd.append('--quiet')
+        if args.zone:
+            cmd.extend(['--zone', args.zone])
+        port_mapping = 'localhost:' + str(args.port) + ':localhost:8080'
+        cmd.extend([
+            '--ssh-flag=-4',
+            '--ssh-flag=-o',
+            '--ssh-flag=LogLevel=error',
+            '--ssh-flag=-N',
+            '--ssh-flag=-L',
+            '--ssh-flag=' + port_mapping])
+        cmd.append('datalab@{0}'.format(instance))
+        with tempfile.TemporaryFile() as tf:
+            gcloud_compute(args, cmd, stdout=tf, stderr=tf)
```

Comment:
Maybe interesting to have a --log file arg that someone can use if they are having issues, and in that case have a verbose output go to that log file. Otherwise, I think tmp file (isn't there a way to use /dev/null instead?) should work just fine.